clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -515,312 +448,319 @@ private void needsCoordinator(FindCoordinatorRequest.CoordinatorType type, Strin
-    }
-
-    private synchronized TransactionalRequest addPartitionsToTransactionRequest(boolean isRetry) {
+    private synchronized TxnRequestHandler addPartitionsToTransactionHandler() {
         pendingPartitionsToBeAddedToTransaction.addAll(newPartitionsToBeAddedToTransaction);
         newPartitionsToBeAddedToTransaction.clear();
         AddPartitionsToTxnRequest.Builder builder = new AddPartitionsToTxnRequest.Builder(transactionalId,
                 pidAndEpoch.producerId, pidAndEpoch.epoch, new ArrayList<>(pendingPartitionsToBeAddedToTransaction));
-        return new TransactionalRequest(builder, new AddPartitionsToTransactionCallback(),
-                FindCoordinatorRequest.CoordinatorType.TRANSACTION, TransactionalRequest.Priority.ADD_PARTITIONS_OR_OFFSETS, isRetry, transactionalId, null);
+        return new AddPartitionsToTxnHandler(builder);
     }
 
-    private TransactionalRequest findCoordinatorRequest(FindCoordinatorRequest.CoordinatorType type, String coordinatorKey, boolean isRetry) {
-        FindCoordinatorRequest.Builder builder = new FindCoordinatorRequest.Builder(type, coordinatorKey);
-        return new TransactionalRequest(builder, new FindCoordinatorCallback(type, coordinatorKey),
-                null, TransactionalRequest.Priority.FIND_COORDINATOR, isRetry, null, null);
-    }
-
-    private TransactionalRequest endTxnRequest(boolean isCommit, boolean isRetry, TransactionalRequestResult result) {
-        EndTxnRequest.Builder builder = new EndTxnRequest.Builder(transactionalId,
-                pidAndEpoch.producerId, pidAndEpoch.epoch, isCommit ? TransactionResult.COMMIT : TransactionResult.ABORT);
-        return new TransactionalRequest(builder, new EndTxnCallback(isCommit, result),
-                FindCoordinatorRequest.CoordinatorType.TRANSACTION, TransactionalRequest.Priority.END_TXN, isRetry, transactionalId, result);
-    }
-
-    private TransactionalRequest addOffsetsToTxnRequest(Map<TopicPartition, OffsetAndMetadata> offsets,
-                                                        String consumerGroupId, boolean isRetry, TransactionalRequestResult result) {
-        AddOffsetsToTxnRequest.Builder builder = new AddOffsetsToTxnRequest.Builder(transactionalId,
-                pidAndEpoch.producerId, pidAndEpoch.epoch, consumerGroupId);
-        return new TransactionalRequest(builder, new AddOffsetsToTxnCallback(offsets, consumerGroupId, result),
-                FindCoordinatorRequest.CoordinatorType.TRANSACTION, TransactionalRequest.Priority.ADD_PARTITIONS_OR_OFFSETS, isRetry, transactionalId, result);
-    }
-
-    private TransactionalRequest txnOffsetCommitRequest(Map<TopicPartition, OffsetAndMetadata> offsets,
-                                                        String consumerGroupId, boolean isRetry, TransactionalRequestResult result) {
+    private TxnRequestHandler txnOffsetCommitHandler(Map<TopicPartition, OffsetAndMetadata> offsets, String consumerGroupId) {
```

Comment:
Ack. Initially the generics helped in a couple cases, but as the refactor progressed, the benefit became marginal. And it is super annoying to add `<?, ?>` everywhere we use these when we don't care about the actual types.